src/gui/src/displayControls.cpp

```diff
@@ -1190,52 +1246,141 @@ bool DisplayControls::isVisible(const odb::dbTechLayer* layer)
+    return true;
+  }
+
   dbMaster* master = inst->getMaster();
   if (master->isEndCap()) {
-    return instances_.endcap.selectable->checkState() == Qt::Checked;
+    return physical_instances_.endcap.selectable->checkState() == Qt::Checked;
   } else if (master->isFiller()) {
-    return instances_.fill.selectable->checkState() == Qt::Checked;
+    return physical_instances_.fill.selectable->checkState() == Qt::Checked;
   } else if (master->isCore()) {
     if (master->getType() == dbMasterType::CORE_WELLTAP) {
-      return instances_.welltap.selectable->checkState() == Qt::Checked;
+      return physical_instances_.welltap.selectable->checkState() == Qt::Checked;
     } else {
-      return instances_.core.selectable->checkState() == Qt::Checked;
+      return physical_instances_.core.selectable->checkState() == Qt::Checked;
     }
   } else if (master->isBlock()) {
-    return instances_.blocks.selectable->checkState() == Qt::Checked;
+    return physical_instances_.blocks.selectable->checkState() == Qt::Checked;
   } else if (master->isPad()) {
-    return instances_.pads.selectable->checkState() == Qt::Checked;
+    return physical_instances_.pads.selectable->checkState() == Qt::Checked;
   } else if (master->isCover()) {
-    return instances_.cover.selectable->checkState() == Qt::Checked;
+    return physical_instances_.cover.selectable->checkState() == Qt::Checked;
   } else {
     return true;
   }
 }
 
+bool DisplayControls::isFunctionalInstanceSelectable(odb::dbInst* inst)
+{
+  if (functional_instance_group_.selectable->checkState() == Qt::Checked) {
+    return true;
+  }
+
+  sta::dbNetwork* network = sta_->getDbNetwork();
+  sta::Cell* cell = network->dbToSta(inst->getMaster());
+  sta::LibertyCell* lib_cell = network->libertyCell(cell);
+  if (lib_cell == nullptr) {
+    return true;
+  }
+  if (lib_cell == nullptr) {
+    return true;
+  }
+  if (lib_cell->isInverter() || lib_cell->isBuffer()) {
+    return functional_instances_.buffer_inv.selectable->checkState() == Qt::Checked;
+  } else if (lib_cell->isClockGate()) {
+    return functional_instances_.clock_gate.selectable->checkState() == Qt::Checked;
+  } if (lib_cell->isLevelShifter()) {
+    return functional_instances_.levelshifter.selectable->checkState() == Qt::Checked;
+  } else if (lib_cell->isPad()) {
+    return functional_instances_.pad.selectable->checkState() == Qt::Checked;
+  } else if (lib_cell->isMacro()) {
+    return functional_instances_.macro.selectable->checkState() == Qt::Checked;
+  } else if (lib_cell->isMemory()) {
+    return functional_instances_.memory.selectable->checkState() == Qt::Checked;
+  } else if (lib_cell->hasSequentials()) {
+    return functional_instances_.sequential.selectable->checkState() == Qt::Checked;
+  } else if (lib_cell->portCount() == 0) {
+    return true; // non-functional
+  } else {
+    // not anything else, so combinational
+    return functional_instances_.combinational.selectable->checkState() == Qt::Checked;
```

Comment:
Rather than repeat this logic in both visible & selectable, could you have a method to return the ModelRow?

Reply:
There are several places where the logic is repeated (nets, and instances) and I thought about putting in a helper function, for those to clean things up a little, so I'll go ahead and do that.